### DISCUSSION TITLE
NavigationView は常に利用する側の View で指定する。

### DISCUSSION THREAD

**Discussion by YusukeHosonuma** [YusukeHosonuma/Effective-SwiftUI#22] [Pending]
## Overview

[NavigationView](https://developer.apple.com/documentation/swiftui/navigationview) の宣言位置は、他コントロールと組み合わせた場合に、とくに厳密に求められることが多く、呼び出し先の View 側に定義すると不都合なケースが多い。（経験則）

また、呼び出し元で指定するようにすれば、NavigationView を利用せずに表示することもできるようになり、再利用性が高まる。

```swift
struct ContentView: View {    
    var body: some View {
        TabView {
            NavigationView { // 🚫 Do not moving inside.
                ChildView()
                    .navigationTitle("Fruits")
                    .toolbar {
                        ToolbarItem {
                            Button(action: {}) {
                                Image(systemName: "square.and.arrow.up")
                            }
                        }
                    }
            }
            .tabItem {
                Image(systemName: "list.bullet")
            }
        }
    }
}

struct ChildView: View {
    var body: some View {
        List(["🍎", "🍌", "🍇"], id: \.self) { item in
            NavigationLink(destination: Text(item)) {
                Text(item)
            }
        }
    }
}
```

## Discussion

現時点では経験則の域を脱しておらず、これが良いプラクティスであるかの確信は持てない。

**Comment by crane-hiromu** [comment]
@YusukeHosonuma 

こちらの意見には概ね賛成なのですが、`TabView`を使う場合は例外かなと思っています🤔

理由としては２点あって、

①
現状、階層が `NavigationView` -> `TabView` だとバグがある


②

`TabView` はトップであるべき的なことをAppleがいっているので（この解釈は難しいですが...）
-> https://developer.apple.com/videos/play/wwdc2022/10001/

<img src="https://user-images.githubusercontent.com/24838521/188847853-b789081f-6206-44f3-ba8b-9ab4ef93913a.png" width=800>


<br>

と言う感じなんですが、いかがでしょうかね🤔
